### ISSUE TITLE
useHydrateAtoms doesn't seem to work after repeated uses on Next.13 [app dir]

### ISSUE THREAD

**Issue by AlanMorel** [pmndrs/jotai#1620]
## Summary
I've noticed that in certain situations inside my Next.13 + app dir project, `useHydrateAtoms` does not seem to actually update the state of the atoms after it is called a few times. I've included a 15 second video of me following the reproduction steps below, in case it is helpful.

### Steps to Reproduce
1. Click reproduction link below and ensure `./data.txt` has `1` inside it. You should see this in console at startup if setup correctly: 
```
Desired value: 1
Jotai value before hydrate: 0
Jotai value after hydrate: 1
```

2. Enter `2` in the text field, hit update. You should see this: 
```
Current database value: 1
Attempting to update to: 2
Database updated to: 2
```

3. Refresh. You should (correctly) see:
```
Desired value: 2
Jotai value before hydrate: 0
Jotai value after hydrate: 2
```

4. Enter `3` in the text field, hit update. You should see: 
```
Current database value: 2
Attempting to update to: 3
Database updated to: 3
```

5. Refresh again. If all went well, you should unfortunately see: 
```
Desired value: 3
Jotai value before hydrate: 2
Jotai value after hydrate: 2
Hydration failed? Desired: 3 vs jotai: 2
```

(If you still don't see the hydration failure, increment the number and refresh again)

The value remains `2` even though I am clearly calling `useHydrateAtoms` with a value of `3`. This is despite the fact that in step 3, it updated as expected.

In other words, I'm not seeing `useHydrateAtoms` work after repeated uses, for some strange reason.

## Link to reproduction

https://stackblitz.com/edit/nextjs-twnmrk?file=data.txt

## Link to video reproduction

https://www.youtube.com/watch?v=HoM2RNzDSic

## Check List

Please do not ask questions in issues.

- [x] I understand this is not a question.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

**Comment by dai-shi** [comment]
Do you use `Provider` in app?
Without `Provider`, it uses a global store in memory which can be shared across session.
`useHydrateAtoms` can set values only if atoms are not initialized (never read).

Otherwise, please let me know if you can reproduce the issue without nextjs. You can create a reproduction based on https://csb.jotai.org .

**Comment by AlanMorel** [comment]
I've added a `Provider` and it appears to have fixed the issue I was having. If I run into another problem, I'll open a new issue, thank you!
